### ISSUE TITLE
RazerS3 producing duplicates

### ISSUE THREAD

**Issue by sarahet** [seqan/seqan#2095]
I used RazerS3 to map 5000 30bp SE reads to hg19 with Hamming distance, 10% error rate and full sensitivity. In the output SAM file I found one mapping doubled although RazerS3 should remove duplicates in the verification step. 

My call:
```
razers3 -fl swift -ng -tc 8 -i 90 -rr 100 -o result.sam hg19.fa reads.fa
```
These are all mappings for a specific read where the 7th and 8th line are the same:
```
2676    0       chr2    28141703        255     30M     *       0       0       GTGCCACCATGCCCGCTCACTTTTTGGTAT  IIIIIIIIIIIIIIIIIIIIIIIIIIIIII  NM:i:3  MD:Z:0A13A2A12
2676    272     chr3    114491029       255     30M     *       0       0       *       *       NM:i:3  MD:Z:4A7A1C15
2676    272     chr4    71942090        255     30M     *       0       0       *       *       NM:i:3  MD:Z:10A1T3T13
2676    272     chr9    79071731        255     30M     *       0       0       *       *       NM:i:3  MD:Z:10A1T1C15
2676    256     chr17   16478688        255     30M     *       0       0       *       *       NM:i:3  MD:Z:17A1T5T4
2676    272     chr17   29879012        255     30M     *       0       0       *       *       NM:i:3  MD:Z:4A5A1T17
2676    272     chr17   63056357        255     30M     *       0       0       *       *       NM:i:3  MD:Z:10T1A0G16
2676    272     chr17   63056357        255     30M     *       0       0       *       *       NM:i:3  MD:Z:10T1A0G16
2676    272     chr17   71191594        255     30M     *       0       0       *       *       NM:i:0  MD:Z:30
```
As mapping position and number, positions and types of mismatches are the same, this should probably be recorded only once or am I missing something?

**Comment by sarahet** [comment]
Edit: This problem seems only to appear in parallel mode.

**Comment by h-2** [comment]
@weese Do you know anything about this?

**Comment by sarahet** [comment]
Just to mention this happened to me also with another dataset with sequences of the same length producing more than 100 duplicates. Do I miss some option that has to be set in order to avoid this?

**Comment by smehringer** [comment]
I just took a quick look so I hope you didn't try that already, but what about setting option `-m 1` ? (or the equivalent `--unique`)
The documentation says:
```
  [ -m NUM ],  [ --max-hits NUM ]

  Output at most NUM of the best matches, default value is 100.

  [ --unique ]

  Output only unique best matches (like ELAND). This flag is equivalent to
  '-m 1 -dr 0 -pa'.
```

**Comment by sarahet** [comment]
As far as I understand it (and also tried) -m controls the maximum number of best matches which means that if I set the sensitivity to 100% - and want all matches I can find with a certain error rate - then this option shows me a limited number (with the 1000 best matches and -m 100 I would only get 100 of this 1000). But that does not mean that I get 100 times the same match at the same position with the exact same mismatches but rather different matches. 
The documentation says also that real duplicates should be filtered out in the verification step.
Does that make sense? I might already thought too much about it ;)

**Comment by sarahet** [comment]
ELAND format would also only give me one record per read but I want the exact positions and mismatches of all possible matches

**Comment by sarahet** [comment]
So I tried it again to make sure that I was correct and indeed setting -m 1 results in a single match output for each read which is not my intention. I do want all different matches that may have the same score but for example different mismatch positions and I assume it should not be the case that the exact same match can be found multiple times in the output.

**Comment by smehringer** [comment]
Hm it seems to be unwanted behaviour then.
Leaves you with ` ... > output | sort | uniq ` until @weese takes a look at it.

**Comment by sarahet** [comment]
Yes that's what I'm doing at the moment :) I just thought it would be good to report it!

**Comment by rrahn** [comment]
@xp3i4 can you have a look please.

**Comment by xp3i4** [comment]
7th and 8th lack enough mapping details. They may be different since they can be mapped in different ways to the same region in the ref that outputs as the same results. not sure.

**Comment by sarahet** [comment]
What do you mean concerning "lack enough mapping details"?  In general they can't be mapped in different ways to the same region as I used the no-gap parameter which means mismatches only. For one read there is only one way it can map to a specific reference region then.

**Comment by xp3i4** [comment]
No gaps in your case seems to be no gaps both for the read and the genome. But if so just searching kmers is enough. It's meaningless to map reads.  
Otherwise it can be mapped in different ways. I think.
so I think may be you can check what -ng exactly mean

**Comment by Stef10x** [comment]
I ran into the exact same issue. It's easy to replicate:
`razers3 -f -o results.razers 3contigs.fa read1.fa`

with 3contigs.fa:
```
>contig1
CAGTGAGTGGAGTGCGAGCTAGAGCGCATCAGCATCAGCATACGAGAAATTGAATGACGT
CTGCAGCCATGGCTTGGACTAGAAAAGAATAATTATTGAATCTGTACTGGCAGGAGATAA
>contig2
TTTAACAGCTAAATATGGTAATGAGTAAAGCACTTGTCTCAGCTTGCGCAGAGCTTGTGC
AGCAGCTTATGCCAGGGAGTCTGAACTGTCTCCATGTACTATAATTAAATTTATAAGTCC
>contig3
CAAGCCCCTCGGCTTCCAGAGCGAACTGAGCGCGCGAGAATGGAGGTCGCCAGCCAAGC
CCTGGGCCCCTTGGTCACCGAGCTGTATGACGTGCAGGCTTTCAAGTTCGGGAGCTTCGTC
```

and read1.fa:
`>read1
CATGGCTTGGACTAGAAAAGAATAA`

returns a duplicate hit:

```
read1	0	25	F	contig1	67	92	100
read1	0	25	F	contig1	67	92	100
```

If you run this on a larger scale the single hit in a single contig is reported several times. Adding `--unique` doesn't do the job returning unique hits. In this example, it results in 0 hits.

This issue seems to get very prominent if you add more contigs. You will see up to 100 reports of the exact same hit.

**Comment by rrahn** [comment]
Hi @Stef10x thanks for writing us. I have to admit that this issue is not easy to track. 
For the time being I recommend trying our read mapper YARA: http://packages.seqan.de/yara/ which is currently maintained, is in general faster and easier to use. If you need assistance with using Yara, please don't hesitate writing us. I can't promise to find time soon to work on this, but we will keep you posted in case something happens here.
